### DISCUSSION TITLE
Is strongswan compatible with NDES ?

### DISCUSSION THREAD

**Discussion by FelixG1316** [strongswan/strongswan#1831] [Q&A]
Hi everyone 👋 

I use strongswan 5.9.11 on a device for SCEP. A customer want to make our SCEP solution with a Microsoft NDES server. Is it possible ?

In such a case, is there a specific configuration on the NDES side to make it work ?

I understood that NDES is like a fork of SCEP made by Microsoft when the RFC 8894 was still in draft state, but I may be wrong 😶 

Thanks for you replies ! 🙏

**Comment by tobiasbrunner** [comment]
> I understood that NDES is like a fork of SCEP made by Microsoft when the RFC 8894 was still in draft state

Well, it was a draft for a very long time (strongSwan's old [`scepclient`](https://docs.strongswan.org/docs/5.9/tools/scepclient.html) implemented one of the drafts 18 years ago). No idea what version NDES might be implementing. So I guess you just have to try if the [SCEP client in `pki`](https://docs.strongswan.org/docs/5.9/pki/pkiScep.html) is compatible.

**Reply by FelixG1316** [reply]
Yes, it was what I concluded. With my team we managed to enroll a certificate with strongswan 5.9.5 and a NDES server, but, for many reasons, we had to upgrade to 5.9.11, and indeed, it is not working anymore with the `pki` CLI.

But now a customer wants to use our solution with its NDES server, and asks the required configuration. We want to answer that is not compatible, but to be sure, we wondered if someone managed to make the last version of strongswan working with NDES, and, in such a case, how.

**Reply by tobiasbrunner** [reply]
What exactly doesn't work? Might just be a question of algorithms or some other minor detail.

**Reply by FelixG1316** [reply]
Unfortunately we do not have more information at the moment, we asked more information to our customer, we are waiting for his answer

**Comment by swer21** [comment]
Hello I used the old scepclient on ubuntu 20.04 to get client certificates from our NDES server. Now when we update to Ubuntu 24.04 strongswan is also updated so I have to trey the new scep Funktion. But I don't have success. 
On the old system I was doing this:
`psec scepclient --out pkcs1=lctest24040.der --out cert=lctest24040.der --dn "C=DE, CN=lctest24040" -s dns=lctest24040.wago.local -s dns=lctest24040 -k 2048 --url "https://svndes01001.wago.local/certsrv/mscep/mscep.dll" --in cacert-enc=caCert-ra-1.der --in cacert-sig=caCert-ra-2.der -p $PASSWORD -f`
With the new client im  doing this.
`ipsec pki --scep --url "https://svndes01001.wago.local/certsrv/mscep/mscep.dll" --password $PASSWORD --outform der`
`ipsec pki --gen > priKey.pem`
`ipsec pki --scep --url "https://svndes01001.wago.local/certsrv/mscep/mscep.dll" --password $PASSWORD --outform der --in priKey.der --dn 'C=DE, CN=lctest24040' --san dns=lctest24040.wago.local -san  dns=lctest24040 --cacert-enc raCert.der --cacert-sig raCert-2.der --cacert caCert.der --cacert caCert-1.der --cipher des3 > lctest2404.der`
But the result is

> TPM 2.0 - could not load "libtss2-tcti-tabrmd.so.0"
> plugin 'tpm': failed to load - tpm_plugin_create returned NULL
> transaction ID: B0362D3B52B1345CCAE23B922909F09EE43EEFDC
> encoding ID_KEY_ID as generalName not supported
> encoding ID_KEY_ID as generalName not supported
>   using trusted certificate "C=DE, ST=NRW, L=Minden, O=WAGO GmbH, OU=IT, CN=SVNDES01001-MSCEP-RA"
>   using trusted intermediate ca certificate "C=DE, DC=local, DC=wago, O=WAGO Kontakttechnik GmbH, CN=WAGO Issuing CA"
>   using trusted ca certificate "C=DE, O=Wago Kontakttechnik GmbH, CN=Wago Root CA"
>   reached self-signed root ca with a path length of 1
> failInfo:     badRequest - transaction not permitted or supported
> reply status is not 'SUCCESS'

Yes I didn't install libtss2-tcti-tabrmd0, because I do not have a TPM chip. But I don't believe that's the issue. 
Im not sure if the "encoding ID_KEY_ID as generalName not supported" is a problem. When I google that nothing helpful to me comes up. 
First I tryed without --cipher des3 but I got a message that aes is not supported.

**Reply by tobiasbrunner** [reply]
> Yes I didn't install libtss2-tcti-tabrmd0, because I do not have a TPM chip. But I don't believe that's the issue.

No, that's not a problem.

> Im not sure if the "encoding ID_KEY_ID as generalName not supported" is a problem.

It's definitely _a_ problem, not sure if it's _the_ problem. It's caused by the SANs configured incorrectly (get rid of the `dns=`).

**Reply by swer21** [reply]
Yes removing the `dns= `part got rid of the ID_KEY_ID massage. But the request is still failing.

> transaction ID: B0362D3B52B1345CCAE23B922909F09EE43EEFDC
  using trusted certificate "C=DE, ST=NRW, L=Minden, O=WAGO GmbH, OU=IT, CN=SVNDES01001-MSCEP-RA"
  using trusted intermediate ca certificate "C=DE, DC=local, DC=wago, O=WAGO Kontakttechnik GmbH, CN=WAGO Issuing CA"
  using trusted ca certificate "C=DE, O=Wago Kontakttechnik GmbH, CN=Wago Root CA"
  reached self-signed root ca with a path length of 1
failInfo:     badRequest - transaction not permitted or supported

**Reply by tobiasbrunner** [reply]
Any logs from the other end? As you probably can imagine, this `badRequest` error code doesn't tell us much.

What you could try is adding `--digest sha1` (the server apparently claims to support the default of `sha256`, otherwise, you'd get an error, but maybe it still doesn't like signatures with it).

**Reply by swer21** [reply]
Unfortunately adding `--digest sha1` dose not help. And I do not have access to the server logs. Not sure if i can get the Admin to cooperate.  If there is nothing that can be done on the client side I may have to try to get a old version of strongswan running under Ubuntu 24.04.

**Reply by tobiasbrunner** [reply]
`scepclient` used DES/MD5 by default, the server seems to at least support 3DES, but maybe it still wants MD5 signatures (I hope not). The latter can't be configured with `pki --scep` (support for DES/MD5 was actually removed once the SCEP RFC was finally released, [which claims](https://datatracker.ietf.org/doc/html/rfc8894#section-7.5) that all known servers support at least 3DES/SHA-1). The old tool also differentiated between the digest used in the signature for the PKCS#10 certificate signing request and the one used for the PKCS#7 signed-data structure in which it is sent to the server, not sure if that could be an issue (it used MD5 for both anyway). I suppose it's always possible the password is wrong or that this particular transaction ID is blocked (it's based on the private key, so you could try generating a new one before changing other request parameters like the digest). Other than that I don't know what differences there might be that could affect this.

**Reply by swer21** [reply]
Unfortunately  the server administrator was not able to tell me what algorithms are allowed for signatures. So I don't know if MD5 is required. But I don't get a error message no matter what value i'm using as digest. And the admin tolled me that the error he has in his log is:
`The password in the certificate request cannot be verified. It may have been used already. Obtain a new password to submit with this request.` 
I'm sure that the password I'm using is valid, even if the massage is disagreeing. I'm getting the password from the website new for each request. And use command substitute to pass it to the ipsec program. I did not include this part for better readability. But I also tryed getting the password manually.

**Reply by tobiasbrunner** [reply]
Yeah, that doesn't sound like an issue with the algorithms. Did you try generating a new private key (which in turn generates a new transaction ID)?

Since the error message is about the password, I checked if there are any differences between 5.8.2 (in Ubuntu 20.04) and current versions in regards to the encoding of the challenge password. And there actually are. Before 5.9.8 (1ef8b922119abc4e22457f731afb0656a8c4ce69) the password was either encoded as ASN.1 PrintableString or ASN.1 T61String (if it contains characters outside of those in PrintableString) as specified by PKCS#9 1.1. Since then the password is always encoded as ASN.1 UTF8String, which is specified in version 2.0 of PKCS#9 (RFC 2985). Although it does say to use PrintableString unless it's necessary to switch to UTF8String. However, it also says: "PKCS#9-attribute processing systems MUST be able to recognize and process all string types". So not sure if that switch to UTF8String could actually be a problem.

**Reply by swer21** [reply]
I was looking into the password encoding but I do not understand if this makes a difference to us. The password is a rundem onetime use password. The password consist only of numbers and uppercase letters. I was under the Impressen that those characters have the same result in those encoding. 
Yes I tested with generating new private keys that did not help. 
In the moment I'm using a old version. Do not believe its a good idea but it works. 
Im using strongswan  only for the scep function, so I may look into a other tool.

**Reply by tobiasbrunner** [reply]
> I was under the Impressen that those characters have the same result in those encoding.

The characters themselves yes, but if the server does not understand the UTF8String encoding at all it might reject the password completely. I suppose we could go back to something similar to what we had before and what the RFC actually says, i.e. use PrintableString unless it's necessary to switch to UTF8String. I can prepare a patch. Would you have time to test that?

**Reply by swer21** [reply]
I have time to test that patch.

**Reply by tobiasbrunner** [reply]
Great, thanks. I've pushed a patch to the _1831-scep-pw-encoding_ branch.

**Reply by swer21** [reply]
Hello, I'm happy to report that I successful requested a certificate from our NDES with the 1831-scep-pw-encoding branch. 

> transaction ID: BB11666142FBB5E35343E09CE8E765DF963A11B9
  using trusted certificate "C=DE, ST=NRW, L=Minden, O=WAGO GmbH, OU=IT, CN=SVNDES01001-MSCEP-RA"
no issuer certificate found for "C=DE, ST=NRW, L=Minden, O=WAGO GmbH, OU=IT, CN=SVNDES01001-MSCEP-RA"
  issuer is "C=DE, DC=local, DC=wago, O=WAGO Kontakttechnik GmbH, CN=WAGO Issuing CA"
  reached end of incomplete trust chain for trusted certificate "C=DE, ST=NRW, L=Minden, O=WAGO GmbH, OU=IT, CN=SVNDES01001-MSCEP-RA"
Issued certificate "C=DE, ST=NRW, L=Minden, O=WAGO GmbH, CN=ubuntu2404stock"
  serial: 51:00:03:2f:85:96:79:24:71:f7:57:e7:e6:00:03:00:03:2f:85
  using certificate "C=DE, ST=NRW, L=Minden, O=WAGO GmbH, CN=ubuntu2404stock"
no issuer certificate found for "C=DE, ST=NRW, L=Minden, O=WAGO GmbH, CN=ubuntu2404stock"
  issuer is "C=DE, DC=local, DC=wago, O=WAGO Kontakttechnik GmbH, CN=WAGO Issuing CA"
Issued certificate is not trusted, valid from May 02 12:15:40 2024 until May 02 12:15:40 2025 (currently valid)

I have to ask the Administrator about the Version of the NDES Server. Maybe there is a update for the server that let it work wit a UTF-8 password. But I expect that the server is up to date.

**Reply by tobiasbrunner** [reply]
Nice, thanks for testing. I'll line the fix up for the next release. The NDES server version would be interesting to know.

**Reply by swer21** [reply]
It turns out that we are still using Windows Server 2016. I did not receive a more precise version number of the NDES server. I have to ask the administrator if he can update it. I would like to know if the problem still exists on a newer server.

**Reply by tobiasbrunner** [reply]
OK, thanks for the information. I wouldn't put it past Microsoft, though, to not have changed anything about this in newer versions.
